src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+                {
+                    var name = nameSyntax.ToString();
+                    var found = false;
+                    for (var j = 0; j < parameters.Length; j++)
+                    {
+                        if (name.Equals(parameters[j].Name))
+                        {
+                            // Check if the argument is in order with parameters.
+                            // If the argument breaks the order, the rest arguments of matched functions must have names
+                            if (i != j)
+                                inOrder = false;
+                            parameterIndex = j;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if (!found) return false;
+                }
+
+                // The argument is either in order with parameters, or have a matched name with parameters
+                var argType = semanticModel.GetTypeInfo(arguments[i].Expression, cancellationToken);
+                if (argType.Type != null && (inOrder || nameSyntax is object))
+                {
+                    // The type of argument must be convertible to the type of parameter
+                    if (!parameters[parameterIndex].IsParams
+                        && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                    {
+                        if (matchedTypes[parameterIndex]) return false;
+                        matchedTypes[parameterIndex] = true;
+                    }
+                    else if (parameters[parameterIndex].IsParams
+                        && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                    {
+                        // The parameter with keyword params takes an array type, then it cannot be matched more than once
+                        if (matchedTypes[parameterIndex]) return false;
+                        matchedTypes[parameterIndex] = true;
+                        paramsMatchedByArray = true;
+                    }
+                    else if (parameters[parameterIndex].IsParams
+                             && parameters.Last().Type is IArrayTypeSymbol paramsType
+                             && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, paramsType.ElementType).Exists)
+                    {
+                        // The parameter with keyword params takes a variable number of arguments, compare its element type with the argument's type.
+                        if (matchedTypes[parameterIndex] && paramsMatchedByArray) return false;
+                        matchedTypes[parameterIndex] = true;
+                        paramsMatchedByArray = false;
+                    }
+                    else return false;
+
+                    if (targetArgument.Equals(arguments[i])) targetParamIndex = parameterIndex;
```

Comment:
please don't do single-line ifs.  they're really unfun to read.  at least wrap (thou you don't  need curlies).